File: backend-model/src/main/java/outcobra/server/model/dto/InstitutionDto.java
package outcobra.server.model.dto;

import outcobra.server.model.User;
import outcobra.server.model.interfaces.*;

public class InstitutionDto implements Identifiable, ParentLinkedDto {
    private final Long id;
    private final String name;
    private final Long userId;

    public InstitutionDto() {
        this.id = this.userId = null;
        this.name = null;
    }

    public InstitutionDto(Long id, String name, Long userId) {
        this.id = id;
        this.name = name;
        this.userId = userId;
    }

    @Override
    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Long getUserId() {
        return userId;
    }

    @Override
    public ParentLink getParentLink() {
        return ParentLink.make(userId, User.class);
    }
}

File: backend/src/main/kotlin/outcobra/server/model/dto/InstitutionDto.kt
package outcobra.server.model.dto

import outcobra.server.model.domain.Institution
import outcobra.server.model.domain.User
import outcobra.server.model.interfaces.OutcobraDto
import outcobra.server.model.interfaces.ParentLink


/**
 * This is a kotlin data class. They need to have a primary constructor.
 * Kotlin will generate instance fields with getters and setters for every parameter within the primary constructor
 * This is a DTO. DTOs are serializable and the only thing that will be sent over the network.
 * In this case it represents an [Institution].
 *
 * For instance field information please check the model class.
 * @see Institution
 */
data class InstitutionDto(val id: Long = 0,
                          val userId: Long = 0,
                          val name: String = "",
                          val schoolClassIds: List<Long> = arrayListOf()) : OutcobraDto {
    /**
     * This function returns the objects unique identifier.
     * This identifier is equal to the primary key of the object in the database
     */
    override val identifier: Long
        get() = id

    /**
     * This function is defined in the [OutcobraDto] interface.
     * This function is used to verify the ownership of an entity.
     */
    override val parentLink: ParentLink
        get() = ParentLink.make(userId, User::class.java)
}